channeldb/kvdb/etcd/stm.go

```diff
@@ -0,0 +1,566 @@
+		if !put.op.IsDelete() && matches(k) {
+			result.key = k
+			result.val = put.val
+			matchFound = true
+		}
+	}
+
+	if !matchFound {
+		return nil
+	}
+
+	return &result
+}
+
+func (s *stm) Next(prefix string, key string) *KV {
+	return s.next(prefix, key, false)
+}
+
+func (s *stm) Seek(prefix, key string) *KV {
+	return s.next(prefix, key, true)
+}
+
+func (s *stm) next(prefix, startKey string, leftClosed bool) *KV {
+	var result KV
+
+	fetchKey := startKey
+	firstFetch := true
+	matchFound := false
+
+	for {
+		opts := []v3.OpOption{
+			v3.WithFromKey(),
+			v3.WithSort(v3.SortByKey, v3.SortAscend),
+			v3.WithLimit(1),
+		}
+
+		if !leftClosed || !firstFetch {
+			fetchKey += "\x00"
+		} else {
+			firstFetch = false
+		}
+
+		kvs := s.fetch(fetchKey, opts...)
+
+		if len(kvs) == 0 {
+			break
+		}
+
+		kv := &kvs[0]
+		// WithRange and WithPrefix can't be used
+		// together, so check prefix here. If the
+		// returned key no longer has the prefix,
+		// then break the fetch loop.
+		if !strings.HasPrefix(kv.key, prefix) {
+			break
+		}
+
+		// Fetch starting with the next key if this
+		// is deleted.
+		if put, ok := s.wset[kv.key]; ok && put.op.IsDelete() {
+			fetchKey = kv.key
+			continue
+		}
+
+		result = *kv
+		matchFound = true
+		break
+	}
+
+	matches := func(k string) bool {
+		if !strings.HasPrefix(k, prefix) {
+			return false
+		}
+
+		if leftClosed && !matchFound {
+			return startKey <= k
+		}
+
+		if !leftClosed && !matchFound {
+			return startKey < k
+		}
+
+		if leftClosed && matchFound {
+			return startKey <= k && k <= result.key
+		}
+
+		// !leftClosed && matchFound
+		return startKey < k && k <= result.key
+	}
+
+	for k, put := range s.wset {
+		if !put.op.IsDelete() && matches(k) {
+			result.key = k
+			result.val = put.val
+			matchFound = true
+		}
+	}
+
+	if !matchFound {
+		return nil
+	}
+
+	return &result
+}
+
+func (s *stm) Put(key, val string, opts ...v3.OpOption) {
+	s.wset[key] = stmPut{
+		val: val,
+		op:  v3.OpPut(key, val, opts...),
+	}
+}
+
+func (s *stm) Del(key string) {
+	s.wset[key] = stmPut{
+		val: "",
+		op:  v3.OpDelete(key),
+	}
+}
+
+func (s *stm) OnCommit(cb func()) {
+	s.onCommit = cb
+}
+
+func (s *stm) Commit() error {
+	cmps := append(s.rset.cmps(), s.wset.cmps(s.revision+1)...)
+	txn := s.client.Txn(s.options.ctx)
+
+	txn = txn.If(cmps...)
+	txn = txn.Then(s.wset.puts()...)
+	// Prefetch keys in case of conflict
+	// to save a round trip to etcd.
+	txn = txn.Else(s.rset.gets()...)
+	txnresp, err := txn.Commit()
+	if err != nil {
+		return DatabaseError{
+			msg: "stm.Commit() failed",
+			err: err,
+		}
+	}
+
+	if txnresp.Succeeded {
+		if s.onCommit != nil {
+			s.onCommit()
+		}
+
+		return nil
+	}
+
+	// Load prefetch to save a round trip.
+	s.rset.add(txnresp)
+	s.prefetch = s.rset
+
+	return CommitError{}
+}
+
+func (s *stm) Rollback() {
+	s.rset = make(map[string]stmGet)
+	s.wset = make(map[string]stmPut)
+	s.getOpts = nil
+	s.revision = math.MaxInt64 - 1
```

Comment:
Not clear on the purpose of this STM level revision.

Reply:
It is the transactions base revision.